### ISSUE TITLE
via fails for PubMed Central full-text articles (eg. http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3245070/ )

### ISSUE THREAD

**Issue by judell** [hypothesis/via#65]
Expect: Annotations

Actual: None

The client query in this case is:

https://hypothes.is/api/search?_separate_replies=true&group=__world__&limit=200&offset=0&order=asc&sort=created&uri=http%3A%2F%2Fpmc%2Farticles%2FPMC3245070%2F

or simply:

https://hypothes.is/api/search?uri=http%3A%2F%2Fpmc%2Farticles%2FPMC3245070%2F

or more readably:

https://hypothes.is/api/search?uri=http://pmc/articles/PMC3245070/ -> 0 results

It should be (and is using the extension or the bookmarklet):

https://hypothes.is/api/search?uri=http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3245070/ -> 10 results

In other words, it appears via has omitted the domain.

**Comment by robertknight** [comment]
TL;DR : The page has a relative `<link rel="canonical">` tag which is not being rewritten by Via on the server side and it should be.

---

This page, as generated by Via, has multiple `<link rel="canonical">` tags, the first is injected by Via and points to the correct URL: http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3245070/ . The second is part of the original page content and it has a relative URL both in the original page and _incorrectly_ after processing by Via: `<link rel="canonical" href="/pmc/articles/PMC3245070/" />`

The href of the link is extracted by Hypothesis using the HTMLLinkElement.href attribute. On `via.hypothes.is` this property is overridden on the client side by a JS function which strips the Via domain from the URL, resulting in a query for http://pmc/articles/PMC3245070/  . In my "lite" version of Via which does not inject this override, the standard href getter is called which resolves the relative URL to http://via.robertknight.me.uk/pmc/articles/PMC3245070/

**Comment by robertknight** [comment]
@nickstenning - You get half marks for predicting this one ;) There is an issue with relative URIs in the `<link>` tag but it isn't related to href attr vs. property.

**Comment by robertknight** [comment]
So by default, pywb enables rewriting of relative canonical URLs (`rewrite_rel_canon` in `config.yaml`). This was explicitly _disabled_ for Via however by 25ed91ebf73ad6a1273715eefa848464ee245a4d . @ikreymer can you remember why rewriting of relative canonical URLs was disabled for Via? It seems to me that this would always produce an incorrect result.

**Comment by robertknight** [comment]
Related Slack discussion: https://hypothes-is.slack.com/archives/public/p1449654928000003

**Comment by robertknight** [comment]
Did some analysis of the most recent 2000 annotated URLs. Of the 1760 HTML pages from that set which I was able to fetch, 701 had `<link rel="canonical">` tags. Of those, there were 16 with relative or root-relative URLs, most of which are PubMed:

```
Annotated URI,Link Tag,Link Rel,Dest
https://www.poets.org/poetsorg/poem/i-hear-america-singing,link,canonical,/poetsorg/poem/i-hear-america-singing
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4418305/,link,canonical,/pmc/articles/PMC4418305/
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1885863/,link,canonical,/pmc/articles/PMC1885863/
https://www.authentichappiness.sas.upenn.edu/newsletters/authentichappinesscoaching/courage,link,canonical,/newsletters/authentichappinesscoaching/courage
https://hackpad.com/A-DigiWriMo-Manifesto-Collaboration-shsym1mvT2b,link,canonical,/A-DigiWriMo-Manifesto-Collaboration-shsym1mvT2b
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4549742/,link,canonical,/pmc/articles/PMC4549742/
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4172799/,link,canonical,/pmc/articles/PMC4172799/
http://www.onthecommons.org/magazine/elinor-ostroms-8-principles-managing-commmons,link,canonical,/magazine/elinor-ostroms-8-principles-managing-commmons
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4064413/,link,canonical,/pmc/articles/PMC4064413/
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4255610/,link,canonical,/pmc/articles/PMC4255610/
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4139440/,link,canonical,/pmc/articles/PMC4139440/
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4099543/,link,canonical,/pmc/articles/PMC4099543/
http://www.grandforksherald.com/news/education/3875306-campus-north-dakota-university-system-continue-support-open-textbook,link,canonical,/news/education/3875306-campus-north-dakota-university-system-continue-support-open-textbook
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3245070/,link,canonical,/pmc/articles/PMC3245070/
http://www.publishersweekly.com/pw/by-topic/industry-news/people/article/54956-e-l-james-pw-s-publishing-person-of-the-year.html,link,canonical,/pw/by-topic/industry-news/people/article/54956-e-l-james-pw-s-publishing-person-of-the-year.html
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4142102/,link,canonical,/pmc/articles/PMC4142102/
```
